File: peek_plugin_diagram/plugin-module/_private/branch/DiagramBranchTuple.ts
import {addTupleType, Tuple} from "@synerty/vortexjs";
import {diagramTuplePrefix} from "../PluginNames";
import {
    BRANCH_DELTA_CLASSES_BY_TYPE,
    DiagramDeltaBase
} from "../../branch/DiagramDeltaBase";
import {ModelSet} from "@peek/peek_plugin_diagram";
import {BranchDeltaBase} from "@peek/peek_plugin_diagram/branch";
import {DiagramCoordSetTuple} from "../..";

/** Diagram Branch Tuple
 *
 * This tuple is used internally to transfer branches from the client tuple provider,
 * and transfer branches to the client wrapped in a TupleAction.
 *
 */
@addTupleType
export class DiagramBranchTuple extends Tuple {
    public static readonly tupleName = diagramTuplePrefix + "DiagramBranchTuple";

    coordSetKey: string;
    coordSetId: number;

    key: string;

    // The list of deltas for this branch
    _packedJson: any[] = [];

    // Properties
    visible: boolean = false;

    private static readonly __COORD_SET_NUM = 0;
    private static readonly __DELTAS_NUM = 1;
    private static readonly __VISIBLE_NUM = 2;

    constructor() {
        super(DiagramBranchTuple.tupleName)
    }

    static unpackJson(key: string, packedJsonStr: string): DiagramBranchTuple {

        // Create the new object
        let newSelf = new DiagramBranchTuple();
        newSelf._packedJson = JSON.parse(packedJsonStr);
        newSelf.key = key;
        return newSelf;

    }

    get coordSetId(): number {
        return this._packedJson[DiagramBranchTuple.__COORD_SET_NUM];
    }

    get deltas(): DiagramDeltaBase[] {
        let deltasJson = this._packedJson[DiagramBranchTuple.__DELTAS_NUM];
        let deltas = [];
        for (let deltaJson of deltasJson) {
            let deltaType = deltaJson[0];
            let Delta = BRANCH_DELTA_CLASSES_BY_TYPE[deltaType];
            let delta = Delta();
            delta._jsonData = deltaJson;
            deltas.push(delta);
        }
        return deltas;
    }

    addDelta(delta: BranchDeltaBase): void {
        this.deltas.push(delta["_deltaJson"])

    }

    get visible(): boolean {
        return this._packedJson[DiagramBranchTuple.__VISIBLE_NUM];
    }
}

File: peek_plugin_diagram/_private/tuples/branch/DiagramBranchTuple.py
import json
from typing import List, Any
from vortex.Tuple import Tuple, addTupleType, TupleField

from peek_plugin_diagram._private.PluginNames import diagramTuplePrefix
from peek_plugin_diagram._private.tuples.branch.BranchDeltaBase import \
    BranchDeltaBase, BRANCH_DELTA_CLASSES_BY_TYPE
from peek_plugin_diagram.tuples.branches.ImportBranchTuple import ImportBranchTuple


@addTupleType
class DiagramBranchTuple(Tuple):
    """ Imported Branch

    This tuple is used by other plugins to load branches into the diagram.

    """
    __COORD_SET_NUM = 0
    __KEY_NUM = 1
    __VISIBLE_NUM = 2

    __tupleType__ = diagramTuplePrefix + 'DiagramBranchTuple'

    #:  The Coordinate Set key that this branch applies to
    coordSetKey: str = TupleField()
    coordSetId: int = TupleField()

    #:  The unique key for this branch
    key: str = TupleField()

    #:  The packed JSON data for this object
    _packedJson:List[Any] = TupleField([])

    #:  Is this branch Visible by default
    visible: bool = TupleField()


    @classmethod
    def packJson(cls, importBranchTuple: ImportBranchTuple, coordSetId: int) -> str:
        """ Pack JSON

        This is used by the import worker to pack this object into the index.

        """

        deltasJson = [
            BranchDeltaBase
        ]
        packedJsonDict = [
            coordSetId,
            deltasJson,
            importBranchTuple.visible
        ]

        return json.dumps(packedJsonDict)

    @@property
    def coordSetId(self):
        return self._packedJson[0]

    @property
    def deltas(self)-> List[BranchDeltaBase]:
        branchDeltasJson = self._packedJson[self.__COORD_SET_ID]
        branchDeltaClasses = []
        for deltaJson in branchDeltasJson:
            branchType = deltaJson[0]
            Delta = BRANCH_DELTA_CLASSES_BY_TYPE[branchType]
            branchDeltaClasses.append(Delta.unpackJson(deltaJson))

        return branchDeltaClasses

    @@property
    def visible(self) -> bool:
        return self._packedJson[self.__VISIBLE_NUM]